core/src/main/java/org/infinispan/remoting/rpc/RpcManager.java

```diff
@@ -21,186 +22,17 @@
-    * @deprecated this method may be removed in the future. Use {@link #invokeRemotelyInFuture(java.util.Collection, org.infinispan.commands.ReplicableCommand, RpcOptions, org.infinispan.commons.util.concurrent.NotifyingNotifiableFuture)}
-    */
-   @Deprecated
-   void invokeRemotelyInFuture(Collection<Address> recipients, ReplicableCommand rpc, boolean usePriorityQueue, NotifyingNotifiableFuture<Object> future);
-
-   /**
-    * The same as {@link #invokeRemotelyInFuture(java.util.Collection, org.infinispan.commands.ReplicableCommand,
-    * boolean, org.infinispan.commons.util.concurrent.NotifyingNotifiableFuture)} except that you can specify a timeout.
-    *
-    * @param recipients       recipients to invoke remote call on
-    * @param rpc              command to execute remotely
-    * @param usePriorityQueue if true, a priority queue is used
-    * @param future           the future which will be passed back to the user
-    * @param timeout          after which to give up (in millis)
-    * @deprecated this method may be removed in the future. Use {@link #invokeRemotelyInFuture(java.util.Collection, org.infinispan.commands.ReplicableCommand, RpcOptions, org.infinispan.commons.util.concurrent.NotifyingNotifiableFuture)}
-    */
-   @Deprecated
-   void invokeRemotelyInFuture(final Collection<Address> recipients, final ReplicableCommand rpc, final boolean usePriorityQueue, final NotifyingNotifiableFuture<Object> future, final long timeout);
-
-   /**
-    * The same as {@link #invokeRemotelyInFuture(java.util.Collection, org.infinispan.commands.ReplicableCommand,
-    * boolean, org.infinispan.commons.util.concurrent.NotifyingNotifiableFuture, long)} except that you can specify a response mode.
-    *
-    * @param recipients       recipients to invoke remote call on
-    * @param rpc              command to execute remotely
-    * @param usePriorityQueue if true, a priority queue is used
-    * @param future           the future which will be passed back to the user
-    * @param timeout          after which to give up (in millis)
-    * @param ignoreLeavers    if {@code true}, recipients that leave or have already left the cluster are ignored
-    *                         if {@code false}, a {@code SuspectException} is thrown when a leave is detected
-    * @deprecated this method may be removed in the future. Use {@link #invokeRemotelyInFuture(java.util.Collection, org.infinispan.commands.ReplicableCommand, RpcOptions, org.infinispan.commons.util.concurrent.NotifyingNotifiableFuture)}
-    */
-   @Deprecated
-   void invokeRemotelyInFuture(Collection<Address> recipients, ReplicableCommand rpc,
-                               boolean usePriorityQueue, NotifyingNotifiableFuture<Object> future,
-                               long timeout, boolean ignoreLeavers);
+   CompletableFuture<Map<Address, Response>> invokeRemotelyAsync(Collection<Address> recipients,
```

Comment:
I don't the use case that well to be able to say, but did you consider defining the return type as `Map<Address, CompletableFuture<Response>>` instead? That might give you more flexibility on the caller side.

Given such a map, the caller can do `CompletableFuture.allOf(map.values())` which returns a CompletableFuture that's completed when all the futures in the map are completed, so that's the effect of your current API.

What's interesting is that you can alternatively do `CompletableFuture.anyOf(map.values())`, and that will give you a CompletableFuture that's completed when any of the given CompletableFutures complete. So, that allows you to wait until one of them is done, giving you the possibility to stop waiting for the rest for example...

I don't know exactly the use case, but in essence, `Map<Address, CompletableFuture<Response>>` gives you more flexibility compared to `CompletableFuture<Map<Address, Response>>`, which can only complete once the Map is completely constructed.